tests/aequilibrae/project/test_field_editor.py

```diff
@@ -1,137 +1,98 @@
+
+
+class TestFieldEditor:
+    @pytest.fixture(scope="session")
+    def project(self, create_empty_project_session):
+        return create_empty_project_session()
+
+    @pytest.fixture(scope="session")
+    def database_backup(self, project):
+        return Path(project.project_base_path).joinpath("project_database.sqlite").read_bytes()
+
+    @pytest.fixture
+    def cleanup_database(self, project, database_backup):
+        """An optional fixture that can be used to revert the project's database in case the test
+        modifies it. This adds some extra overhead to the test (teardown)
+        """
+        yield
+        Path(project.project_base_path).joinpath("project_database.sqlite").write_bytes(database_backup)
+
+    @pytest.fixture(params=["link_types", "links", "modes", "nodes"])
+    def table_name(self, request):
+        return request.param
+
+    @pytest.fixture
+    def table(self, project, table_name):
+        return FieldEditor(project, table_name)
+
+    @pytest.fixture
+    def field_name(self, table):
+        return next(iter(table._original_values.keys()))
+
+    @pytest.fixture
+    def attribute_count(self, table, table_name):
+        qry = f'select count(*) from "attributes_documentation" where name_table="{table_name}"'
+        return table.project.conn.execute(qry).fetchone()[0]
+
+    def test_building(self, table, attribute_count):
+        assert attribute_count == len(table._original_values), "Meta table populated with the wrong number of elements"
+
+    def test_error_when_adding_existing_attribute(self, table, field_name):
+        with pytest.raises(ValueError, match="attribute_name already exists"):
+            table.add(field_name, "some_value")
+
+    @pytest.mark.parametrize(
+        "attribute_name, error",
+        [
+            ("with space", 'attribute_name can only contain letters, numbers and "_"'),
+            ("0starts_with_digit", "attribute_name cannot begin with a digit"),
+        ],
+    )
+    def test_add_invalid_attribute(self, attribute_name, error, table):
+        with pytest.raises(ValueError, match=error):
+            table.add(attribute_name, "some description")
+
+    @pytest.mark.usefixtures("cleanup_database")
+    def test_add_valid_attribute(self, table, table_name, attribute_count):
+        project = table.project
+        new_attribute = "new_attribute"
+        table.add(new_attribute, "some description")
+        project.conn.commit()
+        curr = project.conn.cursor()
+        curr.execute(f'select count(*) from "attributes_documentation" where name_table="{table_name}"')
+        q2 = curr.fetchone()[0]
+        assert q2 == attribute_count + 1, "Adding element did not work"
+
+        result = curr.execute(
+            f'select "{new_attribute}" from "attributes_documentation" where name_table="{table_name}"'
+        ).fetchone()[0]
+        assert result == new_attribute
+
+    @pytest.mark.parametrize(
+        "table_name, attribute, description",
```

Comment:
this is in an interesting use - this table_name isn't consumed in this test, but rather feeds into the "table" fixture which this test consumes. Tricky.

Reply:
Yes, very weird when you see it the first time. I'll be sure to comment it. Or, if preferable, I can refactor and setup this test some other way

Reply:
Maybe a comment - to help anyone who doesn't know pytest that well. No need to refactor - it's working as the framework was designed

Reply:
comment looks perfect.